File: LuyenTai/LuyentapOOP(Kethua).java
//File SQ.java  
public class SQ {
    protected int size , index  ; 
    protected int [] Array ; 
    public SQ(int size){
        this.size = size ; 
        index = -1 ; 
        Array = new int [size];
    }
    public boolean isEmpty(){
        return (index == -1) ;
    }
    public boolean isFull(){
        return index == size - 1;
    }
}
//File stack.java
public class stack extends SQ{
    public stack(int size){
        super(size);
    }
    public void push(int i ){
        if (isFull()){
            System.out.println("Full");
        }else{
            index += 1;
            Array[index] = i;
            System.out.println("Add to :" + i);
        }
    }
    
    public int pop() {
        if (isEmpty()) {
            System.out.println("Stack is Empty");
            return -1;
        } else {
            int value = Array[index];
            index -= 1;
            System.out.println("Popped from stack: " + value);
            return value;
        }
    }
}
//File queue.java
public class queue extends SQ{
    int front , rear ; 
    public queue(int size ){
        super(size);
        this.front = -1;
        this.rear = -1;
    }
    public boolean isFull(){
        return ((front == 0 && rear == size - 1) || (front == rear + 1));
    }
    public void push(int i){
        if (isFull()){
            System.out.println("Full");
        }else{
            if (front == -1 ){
                front = 0;
                rear = (rear + 1) % size;
                Array[rear] = i ;
                System.out.println("Add to :" + i);
            }
        }
    }
    
    public int pop() {
    if (front == -1) {
        System.out.println("Queue is Empty");
        return -1;
    } else {
        int temp = Array[front];
        if (front == rear) {
            // Only one element was present, so reset the queue
            front = -1;
            rear = -1;
        } else {
            // Move front to the next position in the circular manner
            front = (front + 1) % size;
        }
        System.out.println("Popped element is: " + temp);
        return temp;
    }
  }
}
//Main.java
public class Main
{
	public static void main(String[] args) {
		stack s = new stack(2);
		s.push(4);
		s.push(5);
		System.out.println("Is full :" + s.isFull());
		s.pop();
		s.pop();
		System.out.println("Is empty :" + s.isEmpty());
		System.out.println("-------------------------");
		queue q = new queue(5);
		q.push(7);
		q.push(4);
		System.out.println("Is full :" + q.isFull());
		q.pop();
		q.pop();
		System.out.println("Is empty :" + q.isEmpty());
	    
	}
}

File: LuyenTai/LuyentapOOP(Kethua).py
class SQ:
    def __init__(self, size):
        self.size = size
        self.Array = [None] * size
        self.index = -1

    def isEmpty(self):
        return self.index == -1

    def isFull(self):
        return self.index == self.size - 1


class Stack(SQ):
    def __init__(self, size):
        super().__init__(size)

    def push(self, i):
        if self.isFull():
            print("Full")
        else:
            self.index += 1
            self.Array[self.index] = i
            print("Add to: ", i)

    def pop(self):
        if self.isEmpty():
            print("Empty")
            return -1
        else:
            valid = self.Array[self.index]
            self.index -= 1
            print("Pop is: ", valid)
            return valid


class Queue(SQ):
    def __init__(self, size):
        super().__init__(size)
        self.front = -1
        self.rear = -1

    def isFull(self):
        return (self.front == 0 and self.rear == self.size - 1) or (self.front == self.rear + 1)

    def isEmpty(self):
        return self.front == -1

    def push(self, i):
        if self.isFull():
            print("Full")
        else:
            if self.front == -1:
                self.front = 0
            self.rear = (self.rear + 1) % self.size
            self.Array[self.rear] = i
            print("Add to: ", i)

    def pop(self):
        if self.isEmpty():
            print("Empty")
            return -1
        else:
            ele = self.Array[self.front]
            if self.front == self.rear:
                self.front = -1
                self.rear = -1
            else:
                self.front = (self.front + 1) % self.size
            print("Pop is: ", ele)
            return ele


if __name__ == "__main__":
    print("Testing Stack:")
    stack = Stack(3)
    stack.push(1)
    stack.push(2)
    stack.push(3)
    stack.push(4)  # This should print "Full"
    stack.pop()
    stack.pop()
    stack.pop()
    stack.pop()  # This should print "Empty"

    print("\nTesting Queue:")
    queue = Queue(3)
    queue.push(1)
    queue.push(2)
    queue.push(3)
    queue.push(4)  # This should print "Full"
    queue.pop()
    queue.pop()
    queue.pop()
    queue.pop()  # This should print "Empty"
